pool/object.go

```diff
@@ -0,0 +1,110 @@
+// Copyright (c) 2016 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package pool
+
+import (
+	"time"
+
+	"github.com/uber-go/tally"
+)
+
+const (
+	sampleObjectPoolLengthEvery = 100
+)
+
+// TODO(xichen): switch to a native pool based implementation when ready.
+type objectPool struct {
+	opts    ObjectPoolOptions
+	values  chan interface{}
+	alloc   Allocator
+	size    int
+	metrics objectPoolMetrics
+}
+
+type objectPoolMetrics struct {
+	free       tally.Gauge
+	total      tally.Gauge
+	getOnEmpty tally.Counter
+	putOnFull  tally.Counter
+}
+
+// NewObjectPool creates a new pool
+func NewObjectPool(opts ObjectPoolOptions) ObjectPool {
```

Comment:
We should probably move this to m3x as well from m3db.  I know the native pool will be ideal, but in case you do need to use a normal pool too, would be best to use a common one in m3x than a copy/paste into m3aggregator.